### PR TITLE
OAS support for TykOperator

### PR DESCRIPTION
**Pull request by andrei-tyk**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If PRing from your fork, don't come from your `master`!
- [ ] Make sure you are making a pull request against our **`master` branch** (left side). Also, it would be best if you started *your change* off *our latest `master`*.
- [ ] Make sure you are updating [CHANGELOG.md](https://github.com/TykTechnologies/tyk-operator/blob/master/CHANGELOG.md) based on your changes.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] If you've changed API models, please update CRDs.
  - [ ] `make manifests`
  - [ ] `make helm`
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Check your code additions will not fail linting checks:
  - [ ] `gofmt -s -w .`
  - [ ] `go vet ./...`
  - [ ] `golangci-lint run`

### CODE REVIEW

File: api/v1alpha1/tykoasapidefinition_types.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+type OASReference struct {
+	model.Target `json:",inline"`
+	KeyName      string `json:"key_name"`
+}
+
+// TykOASApiDefinitionStatus defines the observed state of TykOASApiDefinition
+type TykOASApiDefinitionStatus struct {
+	ApiID string `json:"api_id"`
+
+	// LinkedByPolicies is a list policies that references this api definition
```
Comment (buraksekili) [inline api/v1alpha1/tykoasapidefinition_types.go:44]:
```suggestion
	// LinkedByPolicies is a list of policy that references this OAS ApiDefinition
```

File: api/v1alpha1/tykoasapidefinition_types.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+type TykOASApiDefinitionSpec struct {
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// Foo is an example field of TykOASApiDefinition. Edit tykoasapidefinition_types.go to remove/update
+	OASRef *OASReference `json:"oas_ref"`
+}
+
+type OASReference struct {
+	model.Target `json:",inline"`
+	KeyName      string `json:"key_name"`
```
Comment (buraksekili) [inline api/v1alpha1/tykoasapidefinition_types.go:37]: can you please add a small comment on what this KeyName refers to?

File: config/samples/oas/petstore_oas_security_policy.yaml [hunk 3]
```diff
@@ -0,0 +1,34 @@
+#apiVersion: tyk.tyk.io/v1alpha1
```
Comment (buraksekili) [inline config/samples/oas/petstore_oas_security_policy.yaml:1]: Why is this sample commented out? isn't valid?

File: controllers/securitypolicy_controller.go [hunk 4]
```diff
@@ -140,33 +140,48 @@ func (r *SecurityPolicyReconciler) spec(
...
 		}
+	}
 
-		r.Log.Error(err, "Failed to get APIDefinition to attach to SecurityPolicy")
+	apiID = api.Status.ApiID
+	apiName = api.Spec.Name
 
-		return err
+	if apiID == "" {
+		r.Log.Info("Failed to get normal APIDefinition to attach to SecurityPolicy, looking for OasApiDefiniiton")
```
Comment (buraksekili) [inline controllers/securitypolicy_controller.go:158]: the same log is duplicated.

File: controllers/securitypolicy_controller.go [hunk 5]
```diff
@@ -140,33 +140,48 @@ func (r *SecurityPolicyReconciler) spec(
...
+
+		apiID = oasApi.Status.ApiID
+		apiName = oasApi.Name
 	}
 
-	if api.Status.ApiID == "" {
+	if apiID == "" {
 		r.Log.Error(
 			opclient.ErrNotFound,
 			"Failed to find ApiDefinition on Tyk", "ApiDefinition", client.ObjectKeyFromObject(api),
```
Comment (buraksekili) [inline controllers/securitypolicy_controller.go:178]: ErrNotFound is happened on cluster resources, not Tyk.

File: controllers/securitypolicy_controller.go [hunk 6]
```diff
@@ -140,33 +140,48 @@ func (r *SecurityPolicyReconciler) spec(
...
 		)
-
 		return opclient.ErrNotFound
```
Comment (buraksekili) [inline controllers/securitypolicy_controller.go:180]: Can't we return early if we couldn't find OAS ApiDef [above](https://github.com/TykTechnologies/tyk-operator/pull/579/files#diff-94d85e85cce9ea50d75136697ae3954462fe96beb26ee1030a7c385cd75ed572R163-R167)?

**Review by buraksekili — COMMENTED** [review]
Also, CI tests are failing. You can skip OAS tests like [this](https://github.com/TykTechnologies/tyk-operator/blob/c285627f70/integration/securitypolicy_test.go#L304-L314).

File: controllers/tykoasapidefinition_controller.go [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+
+	id := tykOASDef.Status.ApiID
+	if id == "" {
+		var result *model.Result
+
+		result, err = klient.Universal.OAS().Create(ctx, data)
+
+		id = result.Meta
+	} else {
+		err = klient.Universal.OAS().Update(ctx, id, data)
```
Comment (buraksekili) [inline controllers/tykoasapidefinition_controller.go:154]: no error check

File: controllers/tykoasapidefinition_controller.go [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+	if data == "" {
+		err = errors.New("OAS Spec is empty")
+		return "", err
+	}
+
+	id := tykOASDef.Status.ApiID
+	if id == "" {
+		var result *model.Result
+
+		result, err = klient.Universal.OAS().Create(ctx, data)
```
Comment (buraksekili) [inline controllers/tykoasapidefinition_controller.go:150]: no error check

File: controllers/tykoasapidefinition_controller.go [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+	} else {
+		err = klient.Universal.OAS().Update(ctx, id, data)
+	}
+
+	if err != nil {
+		r.Log.Error(err, "Failed to create/update OAS API Definition")
+		return "", err
+	}
+
+	r.Log.Info("Successfully created/updated OAS API on Tyk", "id", id)
```
Comment (buraksekili) [inline controllers/tykoasapidefinition_controller.go:162]: maybe, we can print logs after sending Update or Create requests.
So, for example, after creating the OAS, we can print `successfully created OAS on Tyk`.

File: api/v1alpha1/tykoasapidefinition_types.go (hunk 1 above)
Comment (andrei-tyk) [reply api/v1alpha1/tykoasapidefinition_types.go:44]: Fixed

File: api/v1alpha1/tykoasapidefinition_types.go (hunk 2 above)
Comment (andrei-tyk) [reply api/v1alpha1/tykoasapidefinition_types.go:37]: Fixed

File: controllers/securitypolicy_controller.go (hunk 4 above)
Comment (andrei-tyk) [reply controllers/securitypolicy_controller.go:158]: The log is not duplicated as it refers to a different case than the log above.

File: controllers/securitypolicy_controller.go (hunk 6 above)
Comment (andrei-tyk) [reply controllers/securitypolicy_controller.go:180]: **Yes we can, fixed.**

File: controllers/tykoasapidefinition_controller.go (hunk 7 above)
Comment (andrei-tyk) [reply controllers/tykoasapidefinition_controller.go:154]: error is checked and logged below

File: controllers/tykoasapidefinition_controller.go (hunk 8 above)
Comment (andrei-tyk) [reply controllers/tykoasapidefinition_controller.go:150]: error is checked and logged below

File: controllers/tykoasapidefinition_controller.go (hunk 7 above)
Comment (andrei-tyk) [reply controllers/tykoasapidefinition_controller.go:154]: I will move it in the blocks for easier differenciation

File: controllers/tykoasapidefinition_controller.go (hunk 8 above)
Comment (andrei-tyk) [reply controllers/tykoasapidefinition_controller.go:150]: I will move it in the blocks for easier differenciation

File: controllers/tykoasapidefinition_controller.go [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+func (r *TykOASApiDefinitionReconciler) delete(ctx context.Context, tykOASDef *tykv1alpha1.TykOASApiDefinition) error {
+	r.Log.Info("Deleting OAS API on Tyk")
+
+	if util.ContainsFinalizer(tykOASDef, keys.TykOASApiDefinitionFinalizerName) {
+		err := isLinkedToPolicies(ctx, r.Client, r.Log, tykOASDef.Status.LinkedByPolicies)
+		if err != nil {
+			return err
+		}
+
+		err = klient.Universal.OAS().Delete(ctx, tykOASDef.Status.ApiID)
```
[the remark this answers is not in the document]
Comment (andrei-tyk) [reply controllers/tykoasapidefinition_controller.go:176]: It's the same case as for normal apiDefinitions because they use the same structure behind.

**Review by buraksekili — COMMENTED** [review]
LGTM overall, thanks @andrei-tyk. Can you please check linter errors and add simple integration tests?

File: controllers/tykoasapidefinition_controller.go [hunk 11]
```diff
@@ -219,13 +265,16 @@ func (r *TykOASApiDefinitionReconciler) delete(ctx context.Context, tykOASDef *t
 }
 
 func (r *TykOASApiDefinitionReconciler) findOasApiDefDependingOnConfigMap(configMap client.Object) []reconcile.Request {
+	r.Log.Info("Andrei123 entered here1")
```
Comment (buraksekili) [inline controllers/tykoasapidefinition_controller.go:268]: ups

**Review by buraksekili — COMMENTED** [review]
can you please fix CI too? it still fails

**Comment by caroltyk** [comment]
Can we rename the CRD to `OASDefinition`?

**Comment by buraksekili** [comment]
This PR will be used as a reference while implementing OAS Controllers.

**Pull request closed without merging**